LUCENE-10627: Using ByteBuffersDataInput reduce memory copy on compressing data

luyuncheng opened the pull request:
JIRA: https://issues.apache.org/jira/browse/LUCENE-10627
I see When Lucene Do flush and merge store fields, need many memory copies:
```
Lucene Merge Thread #25940]" #906546 daemon prio=5 os_prio=0 cpu=20503.95ms elapsed=68.76s tid=0x00007ee990002c50 nid=0x3aac54 runnable  [0x00007f17718db000]
   java.lang.Thread.State: RUNNABLE
	at org.apache.lucene.store.ByteBuffersDataOutput.toArrayCopy(ByteBuffersDataOutput.java:271)
	at org.apache.lucene.codecs.compressing.CompressingStoredFieldsWriter.flush(CompressingStoredFieldsWriter.java:239)
	at org.apache.lucene.codecs.compressing.CompressingStoredFieldsWriter.finishDocument(CompressingStoredFieldsWriter.java:169)
	at org.apache.lucene.codecs.compressing.CompressingStoredFieldsWriter.merge(CompressingStoredFieldsWriter.java:654)
	at org.apache.lucene.index.SegmentMerger.mergeFields(SegmentMerger.java:228)
	at org.apache.lucene.index.SegmentMerger.merge(SegmentMerger.java:105)
	at org.apache.lucene.index.IndexWriter.mergeMiddle(IndexWriter.java:4760)
	at org.apache.lucene.index.IndexWriter.merge(IndexWriter.java:4364)
	at org.apache.lucene.index.IndexWriter$IndexWriterMergeSource.merge(IndexWriter.java:5923)
	at org.apache.lucene.index.ConcurrentMergeScheduler.doMerge(ConcurrentMergeScheduler.java:624)
	at org.elasticsearch.index.engine.ElasticsearchConcurrentMergeScheduler.doMerge(ElasticsearchConcurrentMergeScheduler.java:100)
	at org.apache.lucene.index.ConcurrentMergeScheduler$MergeThread.run(ConcurrentMergeScheduler.java:682)
```

When Lucene CompressingStoredFieldsWriter do flush documents, it needs many memory copies:

- With Lucene90 using LZ4WithPresetDictCompressionMode:

1. bufferedDocs.toArrayCopy copy blocks into one continue content for chunk compress
2. compressor copy dict and data into one block buffer
3. do compress
4. copy compressed data out

- With Lucene90 using DeflateWithPresetDictCompressionMode:

1. bufferedDocs.toArrayCopy copy blocks into one continue content for chunk compress
2. do compress
3. copy compressed data out

I think we can `use CompositeByteBuf` to **reduce temp memory copies** :
- we do not have to bufferedDocs.toArrayCopy when just need continues content for chunk compress


I write a simple mini benchamrk in test code ([link](https://github.com/apache/lucene/blob/5a406a5c483c7fadaf0e8a5f06732c79ad174d11/lucene/core/src/test/org/apache/lucene/codecs/lucene90/compressing/TestCompressingStoredFieldsFormat.java#L353)): 
LZ4WithPresetDict run Capacity:41943040(bytes) , iter 10times: 
`Origin elapse:5391ms , New elapse:5297ms`
DeflateWithPresetDict run Capacity:41943040(bytes), iter 10times: 
`Origin elapse:115ms, New elapse:12ms`
 
And I run runStoredFieldsBenchmark with doc_limit=-1:
shows:


Msec to index | BEST_SPEED | BEST_COMPRESSION
-- | -- | --
Baseline | 318877.00 | 606288.00
Candidate | 314442.00 | 604719.00

jpountz commented on lucene/core/src/java/org/apache/lucene/codecs/compressing/CompressionMode.java:
```diff
@@ -257,9 +270,13 @@ private static class DeflateCompressor extends Compressor {
 
     @Override
-    public void compress(byte[] bytes, int off, int len, DataOutput out) throws IOException {
+    public void compress(ByteBuffersDataInput buffersInput, int off, int len, DataOutput out)
```
Should we remove `off` and `len` and rely on callers to create a `ByteBuffersDataInput#slice` if they only need to compress a subset of the input?

luyuncheng replied:
it is a nice suggestion, i try to use method `compress(CompositeByteBuf compositeByteBuf, DataOutput out) `

jpountz reviewed (changes requested):
Would it be possible to remove all `CompressionMode#compress` variants that take a `byte[]` now that you introduced a new method that takes a `ByteBuffersDataInput`?

Also maybe we should keep old codecs unmodified and only make this change to `Lucene90Codec` where it makes most sense?

luyuncheng commented:
> Would it be possible to remove all `CompressionMode#compress` variants that take a `byte[]` now that you introduced a new method that takes a `ByteBuffersDataInput`?
> 
> Also maybe we should keep old codecs unmodified and only make this change to `Lucene90Codec` where it makes most sense?

Hi @jpountz Thanks for reviewing this code. 

I prefer keeping old codecs unmodified, because `CompressionMode#compress` is a public abstract method, if we change it with variants `ByteBuffersDataInput` we need to backport in many codecs, like [commits](https://github.com/apache/lucene/blob/382962f22df3ee3af3fb538b877c98d61a622ddb/lucene/backward-codecs/src/java/org/apache/lucene/backward_codecs/lucene50/Lucene50StoredFieldsFormat.java).

And if we only using compress method with variants ByteBuffersDataInput in LUCENE90, we can not using abstract method `Compressor.compress`, when we want to use other compression mode.

Would it be possible to add a new method in Compressor, like following? it can keep the old codecs unmodified, and method with variants ByteBuffersDataInput only can be called in Lucene90Codec.

```
public abstract void compress(byte[] bytes, int off, int len, DataOutput out) throws IOException;

public void compress(CompositeByteBuf compositeByteBuf, DataOutput out) throws IOException;
```

jpountz commented:
> if we only using compress method with variants ByteBuffersDataInput in LUCENE90, we can not using abstract method Compressor.compress, when we want to use other compression mode.

I think that this downside is fine? We prefer codecs to evolve independently so when we start needing changes for a new codec, we prefer to fork the code so that old codecs still rely on the unchanged code (which should move to lucene/backward-codecs) while the new codecs only use the new code without carrying over legacy code.

luyuncheng committed:
1. Using ByteBuffersDataInput reduce stored fields memory copy
2. Using ByteBuffersDataInput reduce TermVectors memory copy
3. Using ByteArrayDataInput reduce memory copy in copyOneDoc
4. Update Compressor.compress method variants with ByteBuffersDataInput
5. Move old Compressor class into backward_codecs

luyuncheng committed:
Tidy

luyuncheng committed:
FIXED len logic

jpountz commented on lucene/core/src/java/org/apache/lucene/codecs/lucene90/compressing/Lucene90CompressingStoredFieldsWriter.java:
```diff
@@ -519,7 +518,13 @@ private void copyOneDoc(Lucene90CompressingStoredFieldsReader reader, int docID)
+      bufferedDocs.copyBytes((ByteArrayDataInput) doc.in, doc.length);
+    } else {
+      bufferedDocs.copyBytes(doc.in, doc.length);
+    }
```
I think that we could avoid this `instanceof` check by overriding `ByteBuffersDataOutput#copyBytes` to read directly into its internal buffers when they are not direct (ie. backed by a `byte[]`)? (Maybe in a separate change?)

luyuncheng replied:
LGTM, i deleted this code at this PR, and i would open a new issue for reduce memory copy in copyOneDoc

luyuncheng replied:
> I think that we could avoid this `instanceof` check by overriding `ByteBuffersDataOutput#copyBytes` to read directly into its internal buffers when they are not direct (ie. backed by a `byte[]`)? (Maybe in a separate change?)

@jpountz  thanks for your advice! i think overriding `ByteBuffersDataOutput#copyBytes` is a great idea, it can reduce many method do memory copy

I opened a new issue: LUCENE-10657 #1034
i think when we move this logic into `ByteBuffersDataOutput#copyBytes` it can reduce more memory copy method:

1. Reduce memory copy in Lucene90CompressingStoredFieldsWriter#copyOneDoc -> bufferdDocs.copyBytes(DataInput input)
2. Reduce memory copy in Lucene90CompoundFormat.writeCompoundFile -> data.copyBytes when input is BufferedChecksumIndexinput and output is ByteBuffersDataOutput
3. Reduce memory IndexWriter#copySegmentAsIs ->CopyFrom -> copyBytes

jpountz commented on lucene/core/src/java/org/apache/lucene/codecs/lucene90/DeflateWithPresetDictCompressionMode.java:
```diff
@@ -199,23 +205,65 @@ private void doCompress(byte[] bytes, int off, int len, DataOutput out) throws I
+        int l = Math.min(blockLength, len - start);
+        // if [start,start + len] stay in one ByteBuffer, we can ignore memory copy
+        // otherwise need to copy bytes into on continuous byte array
+        ByteBuffer bb = buffersInput.sliceOne(start, l);
```
I would prefer abstracting this at the ByteBuffersDataInput level, e.g. maybe there could be a `ByteBuffer ByteBuffersDataInput#readBytes(int numBytes)` API that would automatically return a view of one of its internal buffers when the bytes are fully contained into a single buffer, and create a fresh new ByteBuffer otherwise?

luyuncheng replied:
Nice suggestion, at [commits(ebcbd...)](https://github.com/apache/lucene/blob/ebcbdd05fbf114ca15d04cd4a9172eb51fab9a1a/lucene/core/src/java/org/apache/lucene/store/ByteBuffersDataInput.java#L175) i moved this logic into `ByteBuffersDataInput#readBytes`, in the method `readBytes`, it would determine whether use internal buffers or create a new buffer

jpountz commented on lucene/core/src/java/org/apache/lucene/codecs/lucene90/DeflateWithPresetDictCompressionMode.java:
```diff
@@ -163,12 +165,16 @@ private static class DeflateWithPresetDictCompressor extends Compressor {
+      bufferBlock = BytesRef.EMPTY_BYTES;
     }
 
     private void doCompress(byte[] bytes, int off, int len, DataOutput out) throws IOException {
```
Can we remove this one and require callers to use the ByteBuffer variant instead?

luyuncheng replied:
++ ++，at [commits(ebcbd)](https://github.com/apache/lucene/blob/ebcbdd05fbf114ca15d04cd4a9172eb51fab9a1a/lucene/core/src/java/org/apache/lucene/codecs/lucene90/DeflateWithPresetDictCompressionMode.java) i removed it

jpountz reviewed (commented):
Thanks for the iteration, the high-level idea looks good to me, I left some suggestions.

luyuncheng committed:
1. change statement byteBuffers in Lucene90CompressingStoredFieldsWriter
2. move sliceOne logic into ByteBuffersDataInput level
3. delete copyOneDoc instance ByteArrayDataInput optimize for separate change
4. delete DeflateWithPresetDictCompressionMode#doCompress with param byte[] instead of bytebuffer

luyuncheng committed:
Merge pull request #4 from luyuncheng/Add_CompositeBuf3

move sliceOne logic into ByteBuffersDataInput level

luyuncheng commented on lucene/core/src/java/org/apache/lucene/codecs/lucene90/compressing/Lucene90CompressingStoredFieldsWriter.java:
```diff
@@ -247,21 +249,18 @@ private void flush(boolean force) throws IOException {
       }
     } else {
-      compressor.compress(content, 0, content.length, fieldsStream);
+      ByteBuffersDataInput bytebuffers = bufferedDocs.toDataInput();
```
++，at [commits(ebcbd...)](https://github.com/apache/lucene/blob/ebcbdd05fbf114ca15d04cd4a9172eb51fab9a1a/lucene/core/src/java/org/apache/lucene/codecs/lucene90/compressing/Lucene90CompressingStoredFieldsWriter.java#L249) i fixed it

luyuncheng commented:
> Thanks for the iteration, the high-level idea looks good to me, I left some suggestions.

@jpountz Thanks for your nice suggestions. at latest commits [ebcbdd](https://github.com/apache/lucene/commit/457bab5abfbaa3aabebe99a97e645b4f623be15f)
1. change statement byteBuffers in Lucene90CompressingStoredFieldsWriter
2. move sliceOne logic into ByteBuffersDataInput level and it would determine whether use internal buffers or create a new buffer
3. delete DeflateWithPresetDictCompressionMode#doCompress method which param is byte[]
4. delete copyOneDoc-> readBytes which instance ByteArrayDataInput optimize. and i would create a new issue to reduce memory copy in copyOneDoc

luyuncheng committed:
1. Update boundary check
2. Add changes entry

luyuncheng committed:
1. Update boundary check
2. Add changes entry
3. Remove the deflater hack as LUCENE-9500

luyuncheng committed:
Merge branch 'main' into Add_CompositeBuf

luyuncheng committed:
1. Update boundary check
2. Add changes entry
3. Remove the deflater hack as LUCENE-9500
4. Fetch Upstream And Fixed Conflicts

luyuncheng committed:
1. Add license

luyuncheng committed:
1. Fixed check

luyuncheng committed:
1. Fixed check

jpountz commented on lucene/core/src/java/org/apache/lucene/codecs/lucene90/DeflateWithPresetDictCompressionMode.java:
```diff
@@ -198,22 +200,25 @@ private void doCompress(byte[] bytes, int off, int len, DataOutput out) throws I
     @Override
-    public void compress(byte[] bytes, int off, int len, DataOutput out) throws IOException {
+    public void compress(ByteBuffersDataInput buffersInput, DataOutput out) throws IOException {
+      final int len = (int) (buffersInput.size() - buffersInput.position());
```
Should position() always be 0?

luyuncheng replied:
i think it is right that `position()` always be 0, if caller do compress method with param buffersInput have been sliced. but compress method is a public interface, i think we need check the position for the length

jpountz commented on lucene/core/src/java/org/apache/lucene/store/ByteBuffersDataInput.java:
```diff
@@ -165,6 +165,36 @@ public void readBytes(byte[] arr, int off, int len) throws EOFException {
+   * @param length from position to length
+   * @return ByteBuffer which bytes read from [pos, pos + length]
+   */
+  public ByteBuffer readBytes(int length) throws EOFException {
```
I wonder if it should be called `readNBytes` for consistency with `InputStream#readNBytes`. Maybe @uschindler has thoughts on this and on the implementation.

uschindler replied:
+1, we should rename the method, because it differs from standard readBytes() with multiple params. This one returns a newly allocated ByteBuffer.

luyuncheng replied:
LGTM, rename it with `readNBytes`

jpountz commented on lucene/core/src/test/org/apache/lucene/codecs/lucene90/compressing/TestCompressingStoredFieldsFormat.java:
```diff
@@ -325,4 +341,160 @@ public void testChunkCleanup() throws IOException {
     dir.close();
   }
+
+  public void testSortedSetVariableLengthBigStoredFields() throws Exception {
```
did you add this because the doc values test case found bugs in stored fields?

luyuncheng replied:
In the 1st version, i try to use doc values check compressing and decompressing, and i founded it is same as `TestLucene90DocValuesFormat` testcase. 
at commits [8c59d03](https://github.com/apache/lucene/pull/987/commits/8c59d035c504c9f51c8991a804f73d6ce7074449) i deleted `testSortedSetVariableLengthBigStoredFields`. 

the compressing test case would using `TestDeflateWithPresetDictCompressionMode` and `TestLZ4WithPresetDictCompressionMode`

and i added readNBytes test case in `TestByteBuffersDataInput`

uschindler commented on lucene/core/src/java/org/apache/lucene/store/ByteBuffersDataInput.java:
```diff
@@ -165,6 +165,36 @@ public void readBytes(byte[] arr, int off, int len) throws EOFException {
       }
     }
   }
+  /**
```
here is a newline missing, I wonder why the formatter does not add it?

luyuncheng replied:
Wondering +1, it seems like spotless can not format this situation. and i manually fixed it

luyuncheng committed:
1. Remove len parameter in DeflateWithPresetDictCompressionMode
2. Rename readBytes to readNBytes in ByteBuffersDataInput
3. Delete duplicate testing
4. add readNBytes testing in TestByteBuffersDataInput

luyuncheng commented on lucene/core/src/java/org/apache/lucene/codecs/lucene90/DeflateWithPresetDictCompressionMode.java:
```diff
@@ -163,20 +165,22 @@ private static class DeflateWithPresetDictCompressor extends Compressor {
     }
 
-    private void doCompress(byte[] bytes, int off, int len, DataOutput out) throws IOException {
+    private void doCompress(ByteBuffer bytes, int len, DataOutput out) throws IOException {
```
LGTM, FIXED it

jpountz reviewed (commented):
Thanks, the change looks good to me. Can you also copy unit tests for the code that you copied to lucene/backward-codecs?

Also I wonder if the `readNBytes` trick actually helps in practice, or if we end up crossing ByteBuffer boundaries most of the time in practice, so that it would be better to have a shared byte[] that we read into for every sub-block instead of keeping allocating a new one. Is this something you can check?

luyuncheng commented:
> Also I wonder if the `readNBytes` trick actually helps in practice, or if we end up crossing ByteBuffer boundaries most of the time in practice 

@jpountz  I have the same question that when crossing ByteBuffer would happened. So, i do a simple test:

As the result, i think `have a shared byte[]` is LGTM, i try to update it.

i rewrite the method with readNBytesWithoutCopy and readNBytesWithCopy as follows.

then i try to use arthas `monitor` with commands `monitor -c 10 org.apache.lucene.store.ByteBuffersDataInput readNBytesWith*` AND the running task is runStoredFieldsBenchmark with BEST_COMPRESSION:
it shows:

__#readNBytesWithoutCopy : #readNBytesWithCopy = 1 : 10__ every 10 seconds
![image](https://user-images.githubusercontent.com/12760367/181586795-f5da6a13-b475-43a5-9ade-4c2f52162d6c.png)

codes:
```
  public ByteBuffer readNBytes(int length) throws EOFException {
    .....
    if (block.remaining() >= length) {
      this.pos += length;
      return readNBytesWithoutCopy(block, blockOffset, length);
    } else {
      return readNBytesWithCopy(length);
    }
  }

  private ByteBuffer readNBytesWithoutCopy(ByteBuffer block, int blockOffset, int length) {
    return block.slice(blockOffset, length);
  }

  private ByteBuffer readNBytesWithCopy(int length) throws EOFException {
    ByteBuffer copyBuffer = ByteBuffer.allocate(length);
    readBytes(copyBuffer, length);
    return copyBuffer.rewind().order(ByteOrder.LITTLE_ENDIAN);
  }
```

luyuncheng committed:
1. Copy original compressing unit tests into backward-codec/compressing
2. Use shared byte[[] in ByteBuffersDataInput when readNBytes needs copy

luyuncheng committed:
1. Copy original compressing unit tests into backward-codec/compressing
2. Use shared byte[] in ByteBuffersDataInput when readNBytes needs copy

jpountz commented:
Thanks for running these tests, let's remove `readNBytes` and switch to a shared byte[] that we call `DataInput#readBytes` on instead.

luyuncheng committed:
1. Remove readNBytes From ByteBuffersDataInput

luyuncheng commented:
> Thanks for running these tests, let's remove `readNBytes` and switch to a shared byte[] that we call `DataInput#readBytes` on instead.

@jpountz Thanks a lot for you reviewing for the code. at https://github.com/apache/lucene/pull/987/commits/c0d31d3134653bf6009b798925fa2350e5f7ec9c i removed readNBytes and using shared byte[] buffer in `DeflateWithPresetDictCompressionMode`. 

I try to do more test about why `readNBytes` with a lot of memory copy, i think this is relevant to the:
1. `DeflateWithPresetCompressingCodec` chunk size: `1 << 18`
2. `LZ4WithPresetDictCompressionMode` NUM_SUB_BLOCKS: `10` 
3. `ByteBuffersDataOutput` BITS_PER_BLOCK

The `chunk size` is larger than `block size` which made `ByteBuffersDataInput` fragment always copy datas

jpountz commented on lucene/CHANGES.txt:
```diff
@@ -54,7 +54,8 @@ Improvements
 ---------------------
-(No changes)
+
+* LUCENE-10627: Using ByteBuffersDataInput reduce memory copy on compressing data. (luyuncheng)
```
Move it to the 9.4 section?

luyuncheng replied:
> Move it to the 9.4 section?

DONE

jpountz reviewed (approved):
LGTM.

luyuncheng committed:
1. Update Changes move to 9.4 section

luyuncheng committed:
Merge branch 'main' into Add_CompositeBuf
